src/core/Object3D.js

```diff
@@ -25,6 +25,110 @@ const _zAxis = /*@__PURE__*/ new Vector3( 0, 0, 1 );
 const _addedEvent = { type: 'added' };
 const _removedEvent = { type: 'removed' };
 
+var skipUMWToAvoidLoop = false;
+
+class Object3DMatrixData {
+
+	constructor() {
+
+		this.matrix = new Matrix4();
+		this.matrixWorld = new Matrix4();
+
+		this.matrixAutoUpdate = Object3D.DefaultMatrixAutoUpdate;
+		this.matrixWorldNeedsUpdate = false;
+
+		this.matrixWorldAutoUpdate = Object3D.DefaultMatrixWorldAutoUpdate; // checked by the renderer
+
+		this.parent = null;
+		this.children = [];
+
+		this.position = new Vector3();
+		this.quaternion = new Quaternion();
+		this.scale = new Vector3( 1, 1, 1 );
+
+		// These must exist for all instances, even if unused, to ensure monomorphism.
+		this.updateMatrixWorldBefore = null;
+		this.updateMatrixWorldAfter = null;
+
+	}
+
+	setParent( parent ) {
+
+		this.parent = parent;
+		parent.children.push( this );
+
+	}
+
+	remove( index ) {
+
+		// we can assume that indices match those on the parent object.
+		this.parent.children.splice( index, 1 );
+		this.parent = null;
+
+	}
+
+	updateMatrix() {
+
+		this.matrix.compose( this.position, this.quaternion, this.scale );
+
+		this.matrixWorldNeedsUpdate = true;
+
+	}
+
+	updateMatrixWorld( force ) {
+
+		if ( this.updateMatrixWorldBefore ) {
+
+			this.updateMatrixWorldBefore( force );
+
+		}
+
+		if ( this.matrixAutoUpdate ) this.updateMatrix();
+
+		if ( this.matrixWorldNeedsUpdate || force ) {
+
+			if ( this.parent === null ) {
+
+				this.matrixWorld.copy( this.matrix );
+
+			} else {
+
+				this.matrixWorld.multiplyMatrices( this.parent.matrixWorld, this.matrix );
+
+			}
+
+			this.matrixWorldNeedsUpdate = false;
+
+			force = true;
+
+		}
+
+		// update Children
+
+		const children = this.children;
+
+		for ( let i = 0, l = children.length; i < l; i ++ ) {
+
+			const child = children[ i ];
+
+			if ( child.matrixWorldAutoUpdate === true || force === true ) {
```

Comment:
Wouldn't be faster to use monomorphism even here, and inline the function - i.e. write

Reply:
I believe the current code is correct, and is already monomorphic.  The tree of children beneath an `Object3DMatrixData` exclusively consists of further `Object3DMatrixData` instances.

See:
https://github.com/mrdoob/three.js/blob/6c9205fefdca0003c1b222bd501799d53b1837c3/src/core/Object3D.js#L525

Reply:
Sorry, I did not see that - then it should fine.